### DISCUSSION TITLE
Snapclient setup on Ubuntu

### DISCUSSION THREAD

**Discussion by 01tot10** [snapcast/snapcast#944] [Q&A] [answered]
Dear community!

I'm setting up multi-room audio for the first time using Snapcast, with a Raspberry Pi running Mopidy as a front-end for Snapserver.
The server side is running smoothly following the easy setup procedure explained in the documentation.

I'm having some hickups still on the client side, for which at the moment I'm using my laptop running Ubuntu 20.04 LTS. I thought maybe someone with more experience on the inner workings of Snapcast could help out?

Basically, my trouble is, if I have some program open on my laptop which might be possibly using the audio device for sound output, the audio piped through Snapcast is not able to connect to the sound device, even if there's no audio playing from any other process at the moment.

A typical scenario would be that I have my Firefox open on my computer, try to play some music from Mopidy, but nothing is going through, and snapclient is giving me:
> snapclient[395012]: ALSA lib pcm_dmix.c:1089:(snd_pcm_dmix_open) unable to open slave
> snapclient[395012]: Exception in initAlsa: Can't open default:CARD=PCH, error: Device or resource busy

What seems to work at this point, is to close Firefox, start playing music, and then open up Firefox again.
A similar thing just happened to me when I had Zoom open, and audio piped through Snapcast wouldn't go through.

Knowing, that in general it is possible to have multiple programs outputting sound open on a given computer, I would imagine this is also the case with Snapclient. Could somebody help me out here?

**Comment by kingosticks** [comment, marked as answer]
Only one process may output directly to an ALSA device at a time. You can technically use ALSA dmix to get around that but it doesn't play nicely with snapclient so that's not a good idea here.

However, Ubuntu 20.04 should (I.e. It's the default) be setup so programs output to pulseaudio rather than ALSA directly. And one of the main benefits of going through pulseaudio is that it allows multiple processes to access the sound device. Can you configure snapclient to use pulseaudio instead?

**Reply by 01tot10** [reply]
Hey @kingosticks ! Thanks for your suggestion (and nice to see you here as well)! I ran snapclient just now from command line with pulseaudio as backend, and it does start playing even if I have my browser open. Nice! The relationship between pulseaudio and alsa was something I never really looked into, but this seems like the perfect introduction since it looks like, by definition, alsa "owns" the devices it uses, which leads to our problem in the first place.

I was previously running snapclient as a service, and adding `--player pulse` to `SNAPCLIENT_OPTS=` doesn't seem to be a thing, at least the service crashes immediately when invoked. I'll investigate how to use pulseaudio as a backend when running snapclient as a service, but for now, I can at least use snapclient directly from terminal and it seems to work!

Thanks a lot again!

**Comment by kingosticks** [comment]
> I was previously running snapclient as a service, and adding --player pulse to SNAPCLIENT_OPTS= doesn't seem to be a thing

I would guess the reason that doesn't work is the service runs as user `snapclient` compared to when you run from the command line you run the service as your user. Pulseaudio server runs as a user service (rather than a more traditional system service) and by default won't accept connections from users other than you. You can fix that by following the first part of the instrction at https://docs.mopidy.com/en/latest/running/service/#system-service-and-pulseaudio to edit /etc/pulse/default.pa and then restart.

**Reply by 01tot10** [reply]
> I would guess the reason that doesn't work is the service runs as user `snapclient` compared to when you run from the command line you run the service as your user...

Yes, I think so too, and that's also what my quick search pointed towards. Thanks a lot for that Mopidy-documentation, it does seem to address exactly what I'm facing! I'll do some experimentation when I have a good moment.
